ote/src/clj/ote/services/service_search.clj

```diff
@@ -155,6 +151,83 @@ Negative return value is an invalid match"
+                {}))
+
+(defn- transport-services-page
+  "Queries database for a page of transport services.
+  `db` is the database handle
+  `ids` is a collection of transport service ids.
+  `offset` and `limit` make it possible to use paging instead of fetching all the results"
+  [db ids offset limit]
+  (let [options (if (and offset limit)
+                  {:specql.core/offset offset
+                   :specql.core/limit limit
+                   :specql.core/order-by :ote.db.modification/created
+                   :specql.core/order-direction :desc}
+                  {})]
+    (specql/fetch db ::t-service/transport-service-search-result
+                              search-result-columns
+                              {::t-service/id (op/in ids)}
+                              options)))
+
+(defn- page-of
+  [seq offset limit]
+  (if (and offset limit)
+    (take limit (drop offset seq))
+    seq))
+
+(defn- transport-services-in-operation-area
+  "Queries database for transport services ordered so that best matches to the operation area are first in the results.
+  `db` is the database handle
+  `operation-area` is a collection of operation area names
+  `ids` is a collection of transport service ids.
+  `offset` and `limit` make it possible to use paging instead of fetching all the results"
+  [db ids operation-area offset limit]
+  (let [match-qualities (service-search-match-qualities db ids operation-area)
+        sorted-ids (sort-by match-qualities (keys match-qualities))
+        page (page-of sorted-ids offset limit)
+        results (transport-services db page)]
+    (sort-by
+     (comp match-qualities ::t-service/id)
+     results)))
+
+(defn- without-import-errors [search-result]
+  (update-in search-result
+             [::t-service/external-interface-links]
+             #(mapv hide-import-errors %)))
+
+(defn- without-personal-info
+  "Removes contact information from the `search-result`"
+  [search-result]
+  (dissoc search-result ::t-service/contact-email ::t-service/contact-address ::t-service/contact-phone))
+
+(defn- filtering-operators
+  "Returns a function that removes not interesting companies from the `search-result`"
+  [operators]
+  (fn [search-result]
+    (update search-result ::t-service/service-companies
+            (fn [c]
+              (filter (fn [company]
+                        (when (and operators (::t-service/business-id company))
+                          (.contains operators (::t-service/business-id company))))
+                      c)))))
+(defn- pare-results
+  "Removes information not intended for the frontend from `search-results`
+  Retains `searched-operators` in the search results"
+  [search-results searched-operators]
+  (->> search-results
+       (mapv without-import-errors)
+       (mapv (filtering-operators searched-operators))
+       (mapv without-personal-info)))
+
 (defn search [db {:keys [operation-area sub-type data-content transport-type text operators offset limit]
```

Comment:
Just wondering that when we have more than 10 000 services in production this way of searching services might bite us. Maybe not the thing that we need to change right now, but this is far from optimal.

Reply:
The functionality here hasn't changed. I extracted this stuff into methods in an attempt to clarify what is being done. I suppose it could be possible to filter this stuff out in the queries themselves.